ocr.pm

```diff
@@ -1,31 +1,88 @@
 # Copyright 2009-2013 Bernhard M. Wiedemann
-# Copyright 2012-2020 SUSE LLC
+# Copyright 2012-2023 SUSE LLC
 # SPDX-License-Identifier: GPL-2.0-or-later
 
+=encoding utf8
+
+=head1 OCR Package
+
+Providing optical character recognition (OCR) functionality
+
+=cut
+
 package ocr;
 use Mojo::Base -strict, -signatures;
 require IPC::System::Simple;
+use autodie ':all';
+use File::Temp 'tempdir';
+use Scalar::Util 'blessed';
+use bmwqemu 'fctwarn';
+
+=head2 img_to_str
+
+Abstracts the area of an image to a character string and returns the result.
+
+=head3 Params
+
+=over
+
+=item *
+
+img - A tinycv image
+
+=back
+
+=head3 returns
+
+Character string of characters '-_.:,;()[]{}<>0-9a-zA-Z~+*^@#$%&/=?!§',
+where '§' is placed for each unrecognized character. Implicitly, '§'
+itself can't be recognized. Keep that in mind for comparison!
+Returns undef on failure.
+
+=cut
+
+sub img_to_str ($img) {
+
+    if (blessed($img)) {
+        unless ($img->isa('tinycv::Image')) {
+            fctwarn("Not a reference to object with type tinycv::Image!");
+            return;
+        }
+    }
+    else {
+        fctwarn("Not a reference to object with type tinycv::Image!");
+        return;
+    }
+    my $ret = "";
+    my $tmpdir = tempdir( CLEANUP => 1 );
+    my $img_fpath = $tmpdir . "ocrimg.png";
+    my $stderr_fpath = $tmpdir . "stderr.txt";
+
+    unless ($img->write($img_fpath)) {
+        fctwarn("Writing image to prepare for OCR failed!");
+        return;
+    }
 
-sub tesseract ($img, $area) {
-    my $imgfn = 'ocr.png';
-    my $txtfn = 'ocr';    # tesseract appends .txt automatically o_O
-    my $txt;
+    # -a defines the accuracy threshold per character in percent. GOCR's default
+    # is 95.
+    chomp($ret = `gocr -C '--_.:,;()[]{}<>0-9a-zA-Z~+*^@#\$\%&/=?!' -u § -a 95 $img_fpath 2> $stderr_fpath`);
 
-    if ($area) {
-        $img = $img->copyrect($area->{xpos}, $area->{ypos}, $area->{width}, $area->{height});
+    if ($?) {
+        my $stderr_fh;
+        unless (open($stderr_fh, "<$stderr_fpath")) {
+            fctwarn("Error during OCR occurred but could not read stderr output!");
```

Comment:
```suggestion
            fctwarn('Error during OCR occurred and stderr could not be read');
```

Reply:
Thanks. I have changed this accordingly.